File: src/main/scala/org/rutiger/theatre/actors/exercise/fourth/DunharrowGhost.scala
package org.rutiger.theatre.actors.exercise.fourth

import akka.actor.{Actor, ActorLogging, PoisonPill}
import org.rutiger.theatre.App.Attack
import org.rutiger.theatre.actors.exercise.fourth.DunharrowGhost.GhostSword

import scala.util.Random

class DunharrowGhost extends Actor with ActorLogging {

  private val MAX_SLASHES: Int = 2
  private val r = Random

  override def receive: Receive = {
    case Attack => {
      now % 2 == 0 match {
        case true => sender() ! GhostSword(r.nextInt(MAX_SLASHES) + 1)
        case false => mysteryErrorBox()
      }
    }
    case _ => log.info("bliba bluba blerg!!!! (What are you saying {})", sender().toString())
  }

  def mysteryErrorBox() = {
    now % 3 match {
      case 0 => throw new IllegalStateException("Ghost overflow")
      case 1 => self ! PoisonPill
      case 2 => context.stop(self)
    }
  }

  def now = System.currentTimeMillis()
}

object DunharrowGhost {
  final case class GhostSword(slashes: Int)
}

File: src/main/java/org/rutiger/theatre/actors/exercise/fourth/DunharrowGhost.java
package org.rutiger.theatre.actors.exercise.fourth;

import akka.actor.AbstractActor;
import akka.actor.PoisonPill;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import org.rutiger.theatre.Messages;

import java.util.Random;

/*TODO
Create new warrior, Aragorn. It is an actor router with 3 children
Pickup the policy you want for balancing
Use this actor for create 3 children actor of the new "Warrior"
Use futureAsk pattern
Set up a supervision strategy with your preference in what to do in each case
Battle should respond to the Aragorn actor to keep track
Adapt rest of actor accordingly
 */
public class DunharrowGhost extends AbstractActor {

    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);
    private final Integer MAX_SLASHES = 2;
    private Random r = new Random();

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(Messages.Attack.class, attackMessage -> {
                    if (now() % 2 == 0) {
                        Integer slashes = r.nextInt(MAX_SLASHES) + 1;
                        log.info("slashing {} my ghostly sword", slashes);
                        getSender().tell(new GhostSword(3), getSelf());
                    }
                    else {
                        mysteryErrorBox();
                    }
                })
                .match(Messages.Killed.class, killed -> {
                    log.info("Seems I, {}, killed some enemies, I will tell my liege!!", getSelf().toString());
                    getContext().parent().tell(killed, getSelf());
                })
                .matchAny(anyMsg -> {
                    log.info("bliba bluba blerg!!!! (What are you saying {})", getSender().toString());
                })
                .build();
    }

    private void mysteryErrorBox() {
        Integer errorResult = (int)(now() % 3);
        switch(errorResult) {
            case 0: throw new IllegalStateException("Ghost overflow");
            case 1: getSelf().tell(PoisonPill.getInstance(), getSelf());
            case 2: getContext().stop(getSelf());
        }
    }

    private Long now() { return System.currentTimeMillis(); }

    public class GhostSword {
        private Integer slashes;
        public GhostSword(Integer slashes) {
            this.slashes = slashes;
        }

        public Integer slashes() {
            return slashes;
        }
    }
}
